src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyName.cs

```diff
@@ -303,175 +305,91 @@ public static bool ReferenceMatchesDefinition(AssemblyName? reference, AssemblyN
-            {
-                // allocating or reallocating array by ensuring enough space based on maxCharsToAdd.
-                char[] newresult = new char[destPos + (currentInputPos - prevInputPos) + minReallocateChars];
+                else
+                {
+                    // We have a character we don't want to escape. It's likely there are more, do a vectorized search.
+                    int charsToCopy = stringToEscape.IndexOfAnyExcept(UnreservedReserved);
+                    if (charsToCopy < 0)
+                    {
+                        charsToCopy = stringToEscape.Length;
+                    }
+                    Debug.Assert(charsToCopy > 0);
 
-                if (dest is not null && destPos != 0)
-                    Buffer.BlockCopy(dest, 0, newresult, 0, destPos << 1);
-                dest = newresult;
+                    vsb.Append(stringToEscape.Slice(0, charsToCopy));
+                    stringToEscape = stringToEscape.Slice(charsToCopy);
+                }
             }
-
-            // ensuring we copied everything form the input string left before last escaping
-            while (prevInputPos != currentInputPos)
-                dest[destPos++] = pStr[prevInputPos++];
-            return dest;
-        }
-
-        internal static void EscapeAsciiChar(char ch, char[] to, ref int pos)
-        {
-            to[pos++] = '%';
-            to[pos++] = HexConverter.ToCharUpper(ch >> 4);
-            to[pos++] = HexConverter.ToCharUpper(ch);
         }
 
-        private static bool IsReservedUnreservedOrHash(char c)
+        internal static void PercentEncodeByte(byte ch, ref ValueStringBuilder vsb)
         {
-            if (IsUnreserved(c))
-            {
-                return true;
-            }
-            return RFC3986ReservedMarks.Contains(c);
+            vsb.Append('%');
+            HexConverter.ToCharsBuffer(ch, vsb.AppendSpan(2), 0, HexConverter.Casing.Upper);
         }
 
-        internal static bool IsUnreserved(char c)
-        {
-            if (char.IsAsciiLetterOrDigit(c))
-            {
-                return true;
-            }
-            return RFC3986UnreservedMarks.Contains(c);
-        }
+        [field: AllowNull]
+        private static SearchValues<char> UnreservedReserved => field ??= SearchValues.Create("!#$&'()*+,-./0123456789:;=?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[]_abcdefghijklmnopqrstuvwxyz~");
 
-        internal const char c_DummyChar = (char)0xFFFF;     // An Invalid Unicode character used as a dummy char passed into the parameter
-        private const short c_MaxAsciiCharsReallocate = 40;
-        private const short c_MaxUnicodeCharsReallocate = 40;
-        private const short c_MaxUTF_8BytesPerUnicodeChar = 4;
-        private const short c_EncodedCharsPerByte = 3;
-        private const string RFC3986ReservedMarks = ":/?#[]@!$&'()*+,;=";
-        private const string RFC3986UnreservedMarks = "-._~";
+        private const int StackallocThreshold = 512;
```

Comment:
I assume 512 was selected since that same as `Uri.StackallocThreshold` from the original source.

Reply:
Yes.